### ISSUE TITLE
hyperfs integration

### ISSUE THREAD

**Issue by karissa** [maxogden/dat#407]
after https://github.com/datproject/discussions/issues/26

- main goal of dat is to provide a way to version control and sync code (containers) AND data (files)
- both code and data are represented as files
- some of the files will be linux filesystems (aka containers) that dat can run as a virtual operating system without superuser permissions
- other files may be part of a dataset.  .zip, .csv, .HDF5, .tiff etc
- in the dat CLI there is no difference between a 'container' and a 'file'. dat is essentially a version controlled folder. some files happen to be containers, others happen to be folders, or files
- since everything is a file, we can share common storage abstractions for code and data use cases (as long as we can preserve special file metadata for containers etc)
- a common workflow will be: clone a repository, mount code and data into a container (e.g. "apply" code to the data) to reproduce some computation
- any changes to the container and/or the data should be able to get pushed to another user efficiently (for collaborating on the same set of code + data)

must haves (exact CLI API is TBD, lets just try and agree on workflows for now):
- [ ] 'add' or 'link' or 'track' files in current directory 
- [ ] `dat clone` metadata by default
- [ ] `dat checkout <version>` to get the files from a given version in the local directory. default 'lastest'. 
- [ ] `dat snapshot -m "message"` creates commit and backs it up (good for smaller changes).
- [ ] `dat commit -m "message"` creates metadata commit but doesn't back up (good for large changes that will be pushed soon)
- [ ] `dat push <another-dat-location>` push to a dat somewhere else
- [ ] `dat pull <another-dat-location>` get the graph from the other dat.
- [ ] `dat create <name>`/`dat list` -- do we want to create and list datasets/volumes, just like hyperfs?
- [ ] `dat run <name> <thing>` does essentially hyperfs exec <thing>
- [ ] `dat fetch` to get updated version of metadata

nice to haves:
- [ ] able to unlink or ignore certain files or directories
- [ ] live pushing/pulling/syncing. when something is committed, it will auto-replicate
- [ ]  allow for something like `dat pull foo.com/ubuntu` to work nicely
- [ ] nice archive format (use .zip?) for static storage and manual duplication of dats on the fs

**Comment by karissa** [comment]
## cli UX containery thoughts

init:
```
npm install -g dat
cd ~/myproject
dat init
```

only once, the first time:
```bash
$ dat create myproject http://data.publicbits.org/common/python
$ ls
myproject
$ dat run myproject/ bin/sh 
....done!
myproject $ git clone http://github.com/me/myproject repo
Cloning into 'repo'...
remote: Counting objects: 554, done.
remote: Total 554 (delta 0), reused 0 (delta 0), pack-reused 554
Receiving objects: 100% (554/554), 378.69 KiB | 326.00 KiB/s, done.
Resolving deltas: 100% (308/308), done.
Checking connectivity... done.
myproject $ cd repo && virtualenv env &&  source env/bin/activate && pip install -r requirements.txt
...etc
myproject $ dat snapshot -m "set up dependencies for myproject"
```

doing stuff:
```bash
$ ls
myproject
$ dat run myproject cd repo && source env/bin/activate && python dothethingiwantrightnow.py
i..
am..
doing..the
thing..you..
want..
```

.dat.yml
```yml
- dothethingiwant: 
   - container: myproject
   - run:
       -  myproject cd repo && source env/bin/activate && python dothethingiwantrightnow.py
```

**Comment by maxogden** [comment]
brainstorming new `require('dat')` API:

```js
var Dat = require('dat')
var dat = new Dat({create: false, path: '../', db: memdb, blobs: blobs})
dat.on('error', cb)
dat.graph

var checkout = dat.checkout(hash)
console.log(checkout.id) // uuid
fileRead.pipe(checkout.createWriteStream())
fileRead.pipe(checkout.createWriteStream())
checkout.on('error', log)
checkout.commit({message: 'hi'}, cb)
```

- a module called `dat-fs` would be used to implement what is returned by `dat.checkout`
- we're considering making the object returned by `dat-fs` fully node `fs` API compatible

**Comment by karissa** [comment]
tracking files
```
$ cd newdat
$ dat init
Created a new dat here.
$ ls -a
. .. data.dat
$ cat data.dat/patterns
*.cat
$ dat create myproject
created a new container at myproject.cat
$ ls
data.dat  myproject.cat
$ wget http://someplace.com/path/to/foo.csv
...done!
$ ls 
data.dat myproject.cat foo.csv
$ dat add *.csv
$ cat data.dat/patterns
*.cat
*.csv
$ dat save -m "set up my project and some data"
Now on version ab234ddf2
$ dat status
Nothing has changed.
```

Adding new tracked files
```
$ wget http://someotherplace.com/path/to/bar.csv
$ ls
data.dat myproject.cat somedata.csv bar.csv
$ dat status
MODIFIED bar.csv 
$ dat save -m "added some other new data"
Now on version bd342sdf
$ dat status
Nothing has changed.
$ dat push
Successfully pushed to http://dat.myserver.edu
```

Adding new files that are untracked
```
$ wget http://someotherplace.com/path/to/zoo.json
$ ls  
data.dat myproject.cat foo.csv bar.csv zoo.json
$ dat status
? zoo.json
$ dat add zoo.json
$ dat status
MODIFIED zoo.json
$ dat save -m "added a zoo file"
Now on version bd342sdf
$ dat status
Nothing has changed.
$ dat push
Successfully pushed to http://dat.myserver.edu
```

**Comment by Juul** [comment]
In my ideal world the working tree would always show the files at their current checked out revision in their original format (csv, json, etc) and it would be possible to directly edit the files and do something like:

```
dat commit -a -m "made a change"
```

If someone doesn't want to keep a working tree (waiting for dat to generate the files for each checkout could get tedious for large datasets), then they would have the option of something like a bare git repo, though I'd still want to keep everything in the .dat/ directory for a bare repo and allow people to selectively instead of automatically write the working tree files:

```
dat init --bare # creates .dat/
wget https://example.com/foo.csv 
wget https://example.com/bar.csv 
dat add foo.csv bar.csv
dat commit -a -m "added foo and bar"
echo "adding,a,row" >> foo.csv
dat commit -a -m "added row to foo"
dat checkout <hash_of_first_commit>
ls # now working tree is empty (untracked files/changes would not have been deleted)
dat synctree foo.csv # write foo.csv at current revision
dat synctree # write all files in working tree at current revision
```

Maybe "bare" isn't the right word to use, since this is considerably different from a bare git repo. Also not sure about the right word for the "synctree" command. Another option would be to have the bare-style repo be the default but have a --synctree option for dat checkout like:

```
dat checkout --synctree <hash> 
```
or
```
dat checkout --synctree foo.csv,bar.csv <hash> 
```

In that case it would still be cool to have the option of automatically syncing the working tree on each checkout:

```
dat init --autosync
```

**Comment by maxogden** [comment]
Closing as this is outdated now, follow https://github.com/datproject/discussions/issues/34 for current work
